In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/used-car-price-prediction/Used_Car_Price_Prediction.csv


In [2]:
df = pd.read_csv("/kaggle/input/used-car-price-prediction/Used_Car_Price_Prediction.csv")
df

/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in greater
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.11/dist-packages/pan

car_name  yr_mfr fuel_type  kms_run  sale_price       city  \
0          maruti swift    2015    petrol     8063      386399      noida   
1       maruti alto 800    2016    petrol    23104      265499      noida   
2     hyundai grand i10    2017    petrol    23402      477699      noida   
3          maruti swift    2013    diesel    39124      307999      noida   
4     hyundai grand i10    2015    petrol    22116      361499      noida   
...                 ...     ...       ...      ...         ...        ...   
7395        honda amaze    2018    diesel    53486      604299  ghaziabad   
7396       maruti ignis    2018    petrol     8854      562599    chennai   
7397        honda amaze    2015    petrol    46300      400499       pune   
7398    maruti alto k10    2016    petrol    27245      284099  new delhi   
7399      mahindra thar    2017    diesel    37134      671399  new delhi   

      times_viewed  body_type transmission           variant  ...  \
0            18715  hatchback       manual           lxi opt  ...   
1             2676  hatchback       manual               lxi  ...   
2              609  hatchback       manual   sports 1.2 vtvt  ...   
3             6511  hatchback       manual               vdi  ...   
4             3225  hatchback       manual    magna 1.2 vtvt  ...   
...            ...        ...          ...               ...  ...   
7395          2756      sedan          NaN  1.5 v cvt i-dtec  ...   
7396           640  hatchback       manual     delta 1.2 k12  ...   
7397           795      sedan       manual    1.2 smt i vtec  ...   
7398          1155  hatchback       manual               lxi  ...   
7399          4603        suv       manual    crde 4x4 bs iv  ...   

      total_owners broker_quote original_price  car_rating  \
0                2       397677       404177.0       great   
1                1       272935       354313.0       great   
2                1       469605            NaN       great   
3                1       294262       374326.0       great   
4                1       360716       367216.0       great   
...            ...          ...            ...         ...   
7395             1       630810       787750.0       great   
7396             1       549440            NaN       great   
7397             1       383419            NaN       great   
7398             1       286515       369885.0       great   
7399             1       646565            NaN       great   

                ad_created_on fitness_certificate emi_starts_from  \
0     2021-04-04T07:09:18.583                True            8975   
1     2021-03-22T14:07:32.833                True            6167   
2     2021-03-20T05:36:31.311                True           11096   
3     2021-01-21T12:59:19.299                True            7154   
4     2021-04-01T13:33:40.733                True            8397   
...                       ...                 ...             ...   
7395  2021-02-07T08:05:30.443                True           14036   
7396  2021-03-31T10:21:56.289                True           13068   
7397  2021-03-04T12:40:38.652                True            9303   
7398  2021-03-16T13:31:39.766                True            6599   
7399  2021-03-07T10:10:19.937                True           15595   

     booking_down_pymnt reserved  warranty_avail  
0                 57960    False           False  
1                 39825    False           False  
2                 71655    False           False  
3                 46200    False           False  
4                 54225    False           False  
...                 ...      ...             ...  
7395              90645     True           False  
7396              84390    False           False  
7397              60075     True           False  
7398              42615    False           False  
7399             100710     True           False  

[7400 rows x 29 columns]

  

---

## 📌 Dataset Overview  
Each row in the dataset represents a **unique used car listing** with details such as make, model, year of manufacture, price, ownership, and other listing metadata.  

---

## 📑 Columns Description  

| Column | Description |
|--------|-------------|
| **car_name** | Name of the car (make + model, e.g., *Maruti Swift*, *Hyundai i10*). |
| **yr_mfr** | Year of manufacture of the car. |
| **fuel_type** | Type of fuel used (*Petrol, Diesel, CNG, Electric*, etc.). |
| **kms_run** | Kilometers driven by the car (odometer reading). |
| **sale_price** | Final listed selling price of the car (**target column for regression**). |
| **city** | City where the car is listed. |
| **times_viewed** | Number of times the listing was viewed online. |
| **body_type** | Body type of the car (*Hatchback, Sedan, SUV*, etc.). |
| **transmission** | Gear type (*Manual / Automatic*). |
| **variant** | Variant of the car (e.g., *LXI, VDI, Sports*). |
| **assured_buy** | Whether the car comes with an assured buy option (*True/False*). |
| **registered_city** | City where the car is registered. |
| **registered_state** | State where the car is registered. |
| **is_hot** | Indicates if the car is a “hot” listing (highly demanded). |
| **rto** | Regional Transport Office code of the registration. |
| **source** | Source platform of the listing. |
| **make** | Manufacturer of the car (e.g., *Maruti, Hyundai, Honda*). |
| **model** | Model name of the car (e.g., *Swift, i10, City*). |
| **car_availability** | Availability status of the car (*Available / Sold*). |
| **total_owners** | Number of previous owners (*1, 2, 3*, etc.). |
| **broker_quote** | Price quoted by broker/agent. |
| **original_price** | Original on-road price when the car was new. |
| **car_rating** | Rating of the car’s condition (*Excellent, Good, Fair*, etc.). |
| **ad_created_on** | Date and time when the ad was created. |
| **fitness_certificate** | Whether the car has a valid fitness certificate (*True/False*). |
| **emi_starts_from** | Monthly EMI amount if financed. |
| **booking_down_pymnt** | Minimum down payment required for booking. |
| **reserved** | Whether the car is already reserved (*True/False*). |
| **warranty_avail** | Availability of warranty (*True/False*). |

---  


In [3]:
# LETS UNDERSTAND ABOUT DATASET
print(f"shape of dataset is {df.shape}")
print("\n ======================================================================\n")
print(f"columns present in the dataset are {df.columns}")
print("\n ======================================================================\n")
print(f"columns present in the dataset with there datatypes\n {df.dtypes}")
print("\n ======================================================================\n")
print(f"information about dataset \n {df.info()}")
print("\n ======================================================================\n")
print(f"data discritbtion of every numerical feature \n{df.describe()}")
print("\n ======================================================================\n")
print(f"data describtion of non numerical datatype \n {df.describe(include=['float64','int64'])}")
print("\n ======================================================================\n")
print(f"numercial columns in dataset are {df.select_dtypes(include = ['float','int']).columns.to_list()}")
print("\n ======================================================================\n")
print(f"non numerical columns in dataset are {df.select_dtypes(include = 'O').columns.to_list()}")

shape of dataset is (7400, 29)


columns present in the dataset are Index(['car_name', 'yr_mfr', 'fuel_type', 'kms_run', 'sale_price', 'city',
       'times_viewed', 'body_type', 'transmission', 'variant', 'assured_buy',
       'registered_city', 'registered_state', 'is_hot', 'rto', 'source',
       'make', 'model', 'car_availability', 'total_owners', 'broker_quote',
       'original_price', 'car_rating', 'ad_created_on', 'fitness_certificate',
       'emi_starts_from', 'booking_down_pymnt', 'reserved', 'warranty_avail'],
      dtype='object')


columns present in the dataset with there datatypes
 car_name                object
yr_mfr                   int64
fuel_type               object
kms_run                  int64
sale_price               int64
city                    object
times_viewed             int64
body_type               object
transmission            object
variant                 object
assured_buy               bool
registered_city         object
registered_state       

In [4]:
# LETS CHECK THE NULL VALUES AND DUPLICATES
# LETS DISTRIBUTE COLUMNS WITH ON VALUES DISTRIBUTION FOR EASY VISULIZATION
catcol = []
numcol = []
midcol = []
for col in df.columns:
    if df[col].isna().sum() == 0:
        print(f"THERE NO NULL VALUES PRESENT IN {col}")
        
        # lets check for values
        val = df[col].nunique()
        
        print(f"NUMBER OF UNIQUE VALUES PRESENT IN {col} ARE = {val}")
        if (val >0 and val<12):
            catcol.append(col)
        elif(val>12 and val<50):
            midcol.append(col)
        else:
            numcol.append(col)
        
        print("\n"+"="*80+"\n")
    else:
        print(f"IN {col} HAS NULL VALUES ")
        print(f"NULL VALUES PERCENTAGE {df[col].isna().mean()*100}%")
        print(f"NULL VALUES /TOTAL VALUES = {df[col].isna().sum()}/{df[col].count()}")

        val = df[col].nunique()
        
        print(f"NUMBER OF UNIQUE VALUES PRESENT IN {col} ARE = {val}")
        if (val >0 and val<12):
            catcol.append(col)
        elif(val>12 and val<50):
            midcol.append(col)
        else:
            numcol.append(col)

        print("\n"+"="*80+"\n")

print(f"numercail columns are {numcol}\n")
print(f"categorical columns are {catcol}\n")
print(f"mid values columns are {midcol}\n")


THERE NO NULL VALUES PRESENT IN car_name
NUMBER OF UNIQUE VALUES PRESENT IN car_name ARE = 185


THERE NO NULL VALUES PRESENT IN yr_mfr
NUMBER OF UNIQUE VALUES PRESENT IN yr_mfr ARE = 24


THERE NO NULL VALUES PRESENT IN fuel_type
NUMBER OF UNIQUE VALUES PRESENT IN fuel_type ARE = 5


THERE NO NULL VALUES PRESENT IN kms_run
NUMBER OF UNIQUE VALUES PRESENT IN kms_run ARE = 7176


THERE NO NULL VALUES PRESENT IN sale_price
NUMBER OF UNIQUE VALUES PRESENT IN sale_price ARE = 4519


THERE NO NULL VALUES PRESENT IN city
NUMBER OF UNIQUE VALUES PRESENT IN city ARE = 13


THERE NO NULL VALUES PRESENT IN times_viewed
NUMBER OF UNIQUE VALUES PRESENT IN times_viewed ARE = 3151


IN body_type HAS NULL VALUES 
NULL VALUES PERCENTAGE 1.3918918918918919%
NULL VALUES /TOTAL VALUES = 103/7297
NUMBER OF UNIQUE VALUES PRESENT IN body_type ARE = 5


IN transmission HAS NULL VALUES 
NULL VALUES PERCENTAGE 7.513513513513513%
NULL VALUES /TOTAL VALUES = 556/6844
NUMBER OF UNIQUE VALUES PRESENT IN transmissi

# 🔎 Dataset Observations  

The dataset contains **7400 rows × 29 columns** describing used car listings.  
It includes a mix of **numerical, categorical, boolean, and datetime-like features**.  

---

## 📐 Dataset Shape  
- **Rows:** 7400  
- **Columns:** 29  

---

## 🧾 Data Types Overview  
- **Numerical Columns (9):** `yr_mfr`, `kms_run`, `sale_price`, `times_viewed`, `total_owners`, `broker_quote`, `original_price`, `emi_starts_from`, `booking_down_pymnt`  
- **Categorical Columns (12):** `fuel_type`, `body_type`, `transmission`, `assured_buy`, `is_hot`, `source`, `car_availability`, `car_rating`, `fitness_certificate`, `reserved`, `warranty_avail`, `city/make/state` (mixed)  
- **Boolean Columns (4):** `assured_buy`, `is_hot`, `reserved`, `warranty_avail`  
- **String/Identifier Columns (4+):** `car_name`, `variant`, `model`, `rto`, etc.  
- **Datetime-like Column:** `ad_created_on`  

---

## 🧮 Numerical Feature Insights  

- **Year of Manufacture (`yr_mfr`):**  
  - Range: 1996 – 2021  
  - Avg: ~2014 → most cars are relatively new.  

- **Kilometers Run (`kms_run`):**  
  - Range: 60 – 996,609 km  
  - Avg: ~62,624 km → dataset covers both low and high mileage cars.  

- **Sale Price (`sale_price`):**  
  - Range: ₹0 – ₹38.66 lakh  
  - Avg: ~₹4.55 lakh → most cars are in the affordable/midrange category.  
  - Target column for regression.  

- **Times Viewed (`times_viewed`):**  
  - Range: 3 – 61,930 views  
  - Median: ~1,088 views → listing popularity is highly skewed.  

- **Owners (`total_owners`):**  
  - Range: 1 – 6  
  - Median: 1 → majority are first-owner cars.  

- **Broker Quote vs Sale Price:**  
  - Broker quotes (`broker_quote`) have similar distribution to `sale_price`.  
  - Useful for comparative pricing analysis.  

- **Original Price (`original_price`):**  
  - Available for ~55% of records (missing 44.3%).  
  - Range: ₹96,890 – ₹27.65 lakh.  

- **Loan/EMI-related (`emi_starts_from`, `booking_down_pymnt`):**  
  - Avg EMI: ~₹10,566  
  - Avg Down Payment: ~₹68,233  
  - Indicates financing options are common in listings.  

---

## 🏷️ Categorical Feature Insights  

- **Car Name (`car_name`):**  
  - 185 unique values (make + model).  
  - Rich variety of brands and models.  

- **Fuel Type (`fuel_type`):**  
  - 5 categories → Petrol, Diesel, CNG, Electric, etc.  
  - Balanced distribution expected but likely skewed to Petrol/Diesel.  

- **City (`city`):**  
  - 13 unique cities → dataset covers multiple metro markets.  

- **Body Type (`body_type`):**  
  - 5 categories (Hatchback, Sedan, SUV, etc.).  
  - **~1.4% missing values**.  

- **Transmission (`transmission`):**  
  - Manual/Automatic.  
  - **~7.5% missing values**.  

- **Registered City/State:**  
  - `registered_city`: 243 unique values → wide variety of registrations.  
  - `registered_state`: 16 unique states.  
  - Minimal missing (<0.2%).  

- **Source (`source`):**  
  - 3 platforms.  
  - ~1.7% missing values.  

- **Car Availability (`car_availability`):**  
  - 4 categories (Available, Sold, etc.).  
  - **~8.4% missing values**.  

- **Car Rating (`car_rating`):**  
  - 4 categories (Excellent, Good, Fair, Poor).  
  - Very few missing values (<0.2%).  

- **Fitness Certificate (`fitness_certificate`):**  
  - True/False.  
  - Very few missing (~0.1%).  

- **Boolean Flags:**  
  - `assured_buy`, `is_hot`, `reserved`, `warranty_avail` → clean (no missing).  

---

## 📊 Missing Value Summary  

- **High Missingness:**  
  - `original_price` → **44.3% missing**.  

- **Moderate Missingness:**  
  - `car_availability` → 8.4% missing.  
  - `transmission` → 7.5% missing.  

- **Low Missingness (<2%):**  
  - `body_type`, `source`, `registered_city`, `registered_state`, `car_rating`, `fitness_certificate`, `ad_created_on`.  

- **No Missing:**  
  - Core columns like `car_name`, `yr_mfr`, `fuel_type`, `kms_run`, `sale_price`, `make`, `model`, `broker_quote`.  

---

## ✅ Key Observations  

1. **Target Variable:** `sale_price` → wide range, highly skewed, needs log-scaling.  
2. **Missing Values:** Significant only in `original_price`, moderate in `car_availability` and `transmission`.  
3. **High Cardinality Columns:** `variant` (943), `registered_city` (243), `car_name` (185).  
4. **Potential Feature Leakage:** `broker_quote` is strongly correlated with `sale_price`.  
5. **Temporal Aspect:** `ad_created_on` can be useful for seasonal/demand analysis.  
6. **Boolean Flags (`is_hot`, `assured_buy`, etc.)** are useful for classification tasks.  
7. **Financing Features (`emi_starts_from`, `booking_down_pymnt`)** can help model affordability-related pricing.  

---
